### PR TITLE
Feature/response handler update

### PR DESCRIPTION
**Pull request by denselmann**

### CODE REVIEW

File: app/states/application.ts [hunk 1]
```diff
@@ -1,12 +1,11 @@
...
-import { ApplicationState as ApplicationStateInterface } from "../interfaces";
-
+import { CurrentAnswerTypes, CurrentHandler } from "../../config/handler";
 
 @injectable()
-export class ApplicationState extends BaseState implements ApplicationStateInterface {
-  constructor(@unmanaged() stateSetupSet: State.SetupSet) {
-    super(stateSetupSet);    
+export class ApplicationState extends BaseState<CurrentAnswerTypes, CurrentHandler> {
+  constructor(@unmanaged() setupSet: State.SetupSet<CurrentAnswerTypes, CurrentHandler>) {
```
Comment (antoniusostermann) [inline app/states/application.ts]: Please use our new merged names and our merged state setup set (@Novaal )

File: app/states/application.ts [hunk 2]
```diff
@@ -16,21 +15,21 @@ export class ApplicationState extends BaseState implements ApplicationStateInter
    */
   unhandledIntent(machine: Transitionable) {
     // Although we are in a different state, all i18n conventions ("{state}.{intent}.{platform}.{key}") still apply
-    this.responseFactory.createVoiceResponse().prompt(this.translateHelper.t());
+    this.responseHandler.prompt(this.translateHelper.t());
```
Comment (antoniusostermann) [inline app/states/application.ts]: Please use just `this.prompt(this.t())` to make it even more convenient.

File: app/states/game.ts [hunk 3]
```diff
@@ -14,48 +14,47 @@ import { BackIntentMixin } from "./mixins/backIntent";
...
 export class GameState extends BackIntentMixin(ApplicationState) {
   currentSessionFactory: CurrentSessionFactory;
+  entities: EntityDictionary
 
   constructor(
-    @inject(injectionNames.current.stateSetupSet) stateSetupSet: State.SetupSet,
+    @inject(injectionNames.current.stateSetupSet) stateSetupSet: State.SetupSet<CurrentAnswerTypes, CurrentHandler>,
     @inject(injectionNames.current.sessionFactory) sessionFactory: CurrentSessionFactory,
-    @inject(injectionNames.current.entityDictionary) public entities: EntityDictionary
+    @inject(injectionNames.current.entityDictionary) entities: EntityDictionary
```
Comment (antoniusostermann) [inline app/states/game.ts]: Why did you remove the `public` keyword and assigned it on your own? The `public` way is shorter.

File: app/states/game.ts [hunk 4]
```diff
@@ -14,48 +14,47 @@ import { BackIntentMixin } from "./mixins/backIntent";
...
-  unhandledGenericIntent(machine: Transitionable) {
+   async unhandledGenericIntent(machine: Transitionable) {
     if (this.entities.contains("guessedNumber")) {
       return machine.handleIntent("guessNumber");
     } else if (this.entities.contains("number")) {
       this.entities.set("guessedNumber", this.entities.get("number"));
       return machine.handleIntent("guessNumber");
     } else {
-      return super.unhandledIntent(machine);
+      super.unhandledGenericIntent(machine, "guessedNumber");
```
Comment (antoniusostermann) [inline app/states/game.ts]: Why are you setting the second argument?

File: app/states/main.ts [hunk 5]
```diff
@@ -26,7 +27,7 @@ export class MainState extends ApplicationState {
    * It is called as soon as the application is launched, e. g. if user says "launch xxxxx".
    */
   invokeGenericIntent() {
-    this.responseFactory.createVoiceResponse().prompt(this.translateHelper.t());
+    this.responseHandler.prompt(this.translateHelper.t());
```
Comment (antoniusostermann) [inline app/states/main.ts]: Please use `this.prompt()` and `this.t()`

File: app/states/main.ts [hunk 6]
```diff
@@ -37,18 +38,15 @@ export class MainState extends ApplicationState {
...
-      // Store this number into session and transition to new state      
-      this.currentSessionFactory().set("myNumber", myNumber.toString()),
+    // Store this number into session and transition to new state      
+    await this.currentSessionFactory().set("myNumber", myNumber.toString()),
 
-      // Transition to GameState
-      machine.transitionTo("GameState")
-    ]);
+    // Transition to GameState
+    await machine.transitionTo("GameState")
```
Comment (antoniusostermann) [inline app/states/main.ts]: Why did you change this to not run in parallel anymore?

File: app/states/main.ts [hunk 7]
```diff
@@ -37,18 +38,15 @@ export class MainState extends ApplicationState {
...
 
     // Send response 
-    this.responseFactory.createVoiceResponse().prompt(this.translateHelper.t());
+    this.responseHandler.prompt(this.translateHelper.t());
```
Comment (antoniusostermann) [inline app/states/main.ts]: Change this to `this.prompt(this.t())` ( do it everywhere )

File: index.ts [hunk 8]
```diff
@@ -33,7 +33,7 @@ export function initializeSetups (assistantJs: AssistantJSSetup, stateMachineSet
   assistantJs.registerComponent(genericUtterancesDescriptor);
 
   // Configure components
-  assistantJs.addConfiguration(components);
+  assistantJs.addConfiguration(components as any);
```
Comment (antoniusostermann) [inline index.ts]: ?

File: spec/app/states/game-state.spec.ts [hunk 9]
```diff
@@ -1,98 +1,83 @@
-import { injectionNames, PlatformSpecHelper, MinimalResponseHandler, Session, GenericIntent } from "assistant-source";
+import { injectionNames, Session, GenericIntent } from "assistant-source";
+import { ThisContext } from "../../this-context";
+import { callIntentHelper } from "../../support/call-intent-helper";
+
+
+interface CurrentThisContext extends ThisContext {
+  prepareGuessing(number?);
```
Comment (antoniusostermann) [inline spec/app/states/game-state.spec.ts]: Define a better interface, use variable names, return types, comments.

File: spec/app/states/game-state.spec.ts [hunk 10]
```diff
@@ -1,98 +1,83 @@
...
       
       describe("with a number passed", function() {
-        it("just acts as guessNumberIntent (here: success)", async function(done) {
-          let responseHandler = await this.prepareGuessing(myNumber);
-          expect(responseHandler.endSession).toBeTruthy();
-          expect(this.translateValuesFor("gameState.guessNumberIntent.success")).toContain(responseHandler.voiceMessage);
-          done();
+        it("just acts as guessNumberIntent (here: success)", async function(this: CurrentThisContext) {
+          let responseResult = await this.prepareGuessing(myNumber);
+          expect(await this.translateValuesForGetter()("gameState.guessNumberIntent.success")).toContain(responseResult.voiceMessage!.text);
```
Comment (antoniusostermann) [inline spec/app/states/game-state.spec.ts]: You removed the test for `endSession` everywhere - re-add it everywhere.

File: spec/helpers/setup.js [hunk 11]
```diff
@@ -2,28 +2,28 @@ require("reflect-metadata");
...
   initializeSetups(this.assistantJs, new AssistantSource.StateMachineSetup(this.assistantJs), new AssistantAuthentication.AuthenticationSetup(this.assistantJs));
 
   // Initialize platform specific spec helpers; remember that all of them fulfill the AssistantSource.unifierInterfaces.PlatformSpecHelper interface!
-  this.platforms = {
-    "alexa": new Alexa.SpecHelper(this.specHelper),
-    "googleAssistant": new GoogleAssistant.SpecHelper(this.specHelper)
-  }
-
+  this.alexaSpecHelper = new alexa.AlexaSpecHelper(this.specHelper);
+  this.googleSpecHelper = new google.GoogleSpecHelper(this.specHelper);
```
Comment (antoniusostermann) [inline spec/helpers/setup.js]: The previous approach was better. If you think about running tests on multiple platform, you could create a helper like

```
function runOnPlatform(platformName, ...) {
  this.platforms[platformName].pretendIntentCalled(...);
}
```

You can do such things less easily with your change

File: spec/app/states/game-state.spec.ts [hunk 12]
```diff
@@ -1,98 +1,83 @@
...
 
         // Prepare request
-        responseHandler = await (this.platforms.alexa as PlatformSpecHelper).pretendIntentCalled("guessNumber", false, additionalExtractions);
+        await this.alexaSpecHelper.pretendIntentCalled("guessNumberIntent", false, additionalExtractions);
 
-        // Store myNumber into session
-        currentSessionFactory = this.container.inversifyInstance.get(injectionNames.current.sessionFactory);
-        await currentSessionFactory().set("myNumber", myNumber.toString());
+        // Store number in session factory
+        await this.currentSessionFactoryGetter()().set("myNumber", myNumber.toString());
```
Comment (antoniusostermann) [inline spec/app/states/game-state.spec.ts]: This should have worked in the previous version, too, since you called `pretendIntentCalled` before

File: spec/app/states/main-state.spec.ts [hunk 13]
```diff
@@ -1,79 +1,37 @@
...
-        expect(parseInt(myNumber)).toBeGreaterThan(0);
-        expect(parseInt(myNumber)).toBeLessThanOrEqual(10);
-        done();
-      });
-
-      it("sets the current conversation state to 'GameState'", async function(done) {
-        currentStateNameProvider = this.container.inversifyInstance.get(injectionNames.current.stateNameProvider);
-        let currentStateName = await currentStateNameProvider();
-        expect(currentStateName).toEqual("GameState");
-        done();
```
Comment (antoniusostermann) [inline spec/app/states/main-state.spec.ts]: Why did you remove all these useful specs and replaced them with a unfitting one?

File: spec/helpers/setup.js [hunk 14]
```diff
@@ -2,28 +2,28 @@ require("reflect-metadata");
...
+  
   // Execute bindings
   this.specHelper.prepare();
 
   // Shorten access to i18next helper
-  this.translateValuesFor = this.container.inversifyInstance.get(AssistantSource.injectionNames.i18nTranslateValuesFor);
+  this.translateValuesForGetter = () => this.container.inversifyInstance.get(AssistantSource.injectionNames.current.i18nTranslateValuesFor);
+
+  // Session Factory
+  this.currentSessionFactoryGetter = () => this.container.inversifyInstance.get(AssistantSource.injectionNames.current.sessionFactory);
```
Comment (antoniusostermann) [inline spec/helpers/setup.js]: You shouldn't need this in any of your specs

File: app/states/game.ts (hunk 4 above)
Comment (denselmann) [reply app/states/game.ts]: Because the method expects at least two arguments

File: app/states/game.ts [hunk 15]
```diff
@@ -36,9 +34,15 @@ export class GameState extends BackIntentMixin(ApplicationState) {
   @needs("guessedNumber")
   public async guessNumberIntent() {
     // Retrieve myNumber from session and given number from entity dictionary
-    let guessedNumber = this.entities.get("guessedNumber");
+    let guessedNumber = this.entities.get("guessedNumber") || "0";
+    // Retrieve calculated number from session factory
     let myNumber = await this.currentSessionFactory().get("myNumber") || "";
 
+    // Unfortunately, there is a bug in the current fakeredis 2.0.0 version so we have to do a typecast
+    if(typeof(guessedNumber === "number")) {
```
Comment (antoniusostermann) [inline app/states/game.ts]: Do you really need this? `fakeredis` shouldn't have any impact on `guessedNumber` because it is an entity, nothing coming from `redis`. I guess you added this because you chagned your spec to pass `myNumber` as number, not as string (see comment below).

File: app/states/game.ts [hunk 16]
```diff
@@ -36,9 +34,15 @@ export class GameState extends BackIntentMixin(ApplicationState) {
   @needs("guessedNumber")
   public async guessNumberIntent() {
     // Retrieve myNumber from session and given number from entity dictionary
-    let guessedNumber = this.entities.get("guessedNumber");
+    let guessedNumber = this.entities.get("guessedNumber") || "0";
```
Comment (antoniusostermann) [inline app/states/game.ts]: You don't need this `|| "0"`

File: spec/app/states/game-state.spec.ts [hunk 17]
```diff
@@ -1,14 +1,15 @@
...
-  prepareGuessing(number?);
+  /** Invokes the guessNumberIntent and returns specific response results */
+  prepareGuessing(guessedNumber?: number | undefined): Promise<Partial<BasicAnswerTypes>>;
+  /** Simulate an intent call and returns the response results */
+  callIntent(intent: GenericIntent | string): Promise<Partial<BasicAnswerTypes>>;
 }
 
 describe("GameState", function() {
-  const myNumber = "2";
+  const myNumber = 2;
```
Comment (antoniusostermann) [inline spec/app/states/game-state.spec.ts]: If you set this to a string (like it is in the real world), you shouldn't have any problems..
